File: algo/lc.53.cc
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        if (nums.empty()) return 0;
        
		// DP: g(i) = max{ a_i, g(i-1) + a_i }
		//     f(i) = max_{j=1}^i g(j)
        vector<int> f(nums.size(), 0);
        f[0] = nums[0];
        for (int i = 1; i < nums.size(); i++) {
            f[i] = max(f[i-1]+nums[i], nums[i]);
        }
        int max = INT_MIN;
        for (int i : f) max = (i > max) ? i : max;
        return max;
    }
};

File: algo/lc.53.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        if not nums:
            return 0
        elif len(nums) == 1:
            return nums[0]
        else:
            dp = [nums[0]]
            for (i, x) in enumerate(nums):
                if i == 0:
                    continue
                # dp[i] = max{x_i, x_1+dp[i-1]}
                val_include = x + dp[-1]
                val_exclude = x
                dp.append(max(val_include, val_exclude))
            return max(dp)
            
